### DISCUSSION TITLE
Chat history with dspy

### DISCUSSION THREAD

**Discussion by MeirKaD** [stanfordnlp/dspy#917] [Q&A]
I am working with my own rm client, I use FAISS & google vertex Ai embedding model to create a chat RAG I am wondering if there is an easy setup for chat history like lang chain's chat history for example ? I don't care implementing this myself it's not hard but maybe there's something I am not familiar with that can make my life easier ? 

Thanks in advance!

**Comment by MarkusOdenthal** [comment]
I implemented LangSmith for my project. The only problem is that I can only trace input/output in my Modules I wrote by myself. Would this be enough for your usecase?

**Comment by kjoth** [comment]
@MeirKaD 

1. You mentioned that you created your own RM client. Could you please share the code for reference? It would be really helpful.
2. Were you able to find a solution for the history issue?

**Comment by MukundaKatta** [comment]
DSPy 2.5+ has a built-in `History` type. Declare `history: dspy.History = dspy.InputField()` in your signature, pass a `dspy.History(messages=[])` when calling, and append `{question, answer}` each turn. The adapter handles rendering it into the LM's chat format. Persist `history.messages` to wherever you like (Redis/SQLite/disk), keyed by user id for multi-user.
